torch/_dynamo/backends/debugging.py

```diff
@@ -18,6 +20,58 @@ def eager(gm, fake_tensor_inputs):
     return gm
 
 
+@register_backend(name="explain")
+def explain(gm: torch.fx.GraphModule, example_inputs):
+    """
+    Explain `compiler_fn` that provide explanation for graph breaks and logs modules' graph information.
+    This function prints the graph structure if `config.explain_print_graphs` is set to True.
+    It uses the `tabulate` library to display the node information in a table format.
+    If `tabulate` is not installed, it prints an error message suggesting to install it using `pip install tabulate`.
+
+    In case of any graph break while compilation,
+    it prints the break reason and the formatted stack trace for debugging-purposes.
+
+    Args:
+        gm (torch.fx.GraphModule): The module for debugging.
+        example_inputs: Inputs to the module.
+
+    Returns:
+        gm.forward as callable.
+
+    Example usage:
+        ```
+        opt_fn = torch.compile(fn,backend="explain") #  or torch._dynamo.optimize("explain")(fn)
+        ```
+    """
+    if config.explain_print_graphs:
+        try:
+            from tabulate import tabulate
+        except ImportError:
+            print(
+                "`Explain Backend` relies on the library `tabulate`, "
+                "which could not be found on this machine. Run `pip "
+                "install tabulate` to install the library."
+            )
+
+        node_specs = [
+            [n.op, n.name, n.target, n.args, n.kwargs] for n in gm.graph.nodes
+        ]
+        print(
+            "Node:\n",
+            tabulate(
+                node_specs, headers=["opcode", "name", "target", "args", "kwargs"]
+            ),
+        )
+    if gm.compile_subgraph_reason.graph_break:
+        formatted_stack = "".join(
+            traceback.format_list(gm.compile_subgraph_reason.user_stack)
+        )
+        msg = f"{gm.compile_subgraph_reason.reason}\n{formatted_stack}"
+        print("Graph Break\nBreak reason:\n", msg)
```

Comment:
Nit but would be prettier to have something like `Break Reason {i}:`

Reply:
Output is updated with break reason number.
```
...
Graph Break
Break reason 4:
 call_function graph_break in skip_files ~/pytorch/torch/_dynamo/__init__.py
  File "~/pytorch/main3.py", line 15, in fn3
    torch._dynamo.graph_break()
```